### PR TITLE
Add parameter resolution protocol

### PR DESCRIPTION
**Pull request by dstrain115**
- Add cirq.is_parameterized() and cirq.resolve_parameters() in Protocols
    - Add _is_parameterized_() and _resolve_parameters_() magic methods to
    several gate and operation classes
    - Remove ParameterizableEffect and its references and a few references
    to extensions.

### CODE REVIEW

File: cirq/protocols/resolve_parameters.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+# limitations under the License.
+
+from typing import Any, TypeVar
+from typing_extensions import Protocol
+from cirq.study import ParamResolver
+
+TDefault = TypeVar('TDefault')
+
+class SupportsParameterization(Protocol):
+    """An object that may be describable by a unitary matrix."""
```
Comment (kevinsung) [inline cirq/protocols/resolve_parameters.py]: Incorrect doc.

Comment (dstrain115) [reply cirq/protocols/resolve_parameters.py]: Uh oh, it appears he has discovered my main programming tactic (cut and paste).
Fixed.

File: cirq/circuits/circuit.py [hunk 2]
```diff
@@ -1126,13 +1127,11 @@ def to_text_diagram_drawer(
...
         return diagram
 
-    def is_parameterized(self,
-                         ext: extension.Extensions = None) -> bool:
-        if ext is None:
-            ext = extension.Extensions()
-        return any(cast(ops.ParameterizableEffect, op).is_parameterized()
-                   for op in self.all_operations()
-                   if ext.try_cast(ops.ParameterizableEffect, op) is not None)
+    def is_parameterized(self) -> bool:
```
Comment (Strilanc) [inline cirq/circuits/circuit.py]: `return any(cirq.is_parameterized(op) for op in self.all_operations())`

File: cirq/ops/common_gates.py [hunk 3]
```diff
@@ -18,6 +18,7 @@
 )
 
 import numpy as np
+import cirq
```
Comment (Strilanc) [inline cirq/ops/common_gates.py]: Do not import `cirq` from within `cirq`. Use `from cirq import protocols` to get at `is_parameterized`.

File: cirq/ops/controlled_gate.py [hunk 4]
```diff
@@ -15,6 +15,7 @@
 from typing import Optional, TypeVar, Type, cast, Union, Sequence
 
 import numpy as np
+import cirq
```
Comment (Strilanc) [inline cirq/ops/controlled_gate.py]: `from cirq import protocols`

File: cirq/protocols/resolve_parameters.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        resolution.  Returns True if the gate has any unresolved Symbols
+        and False otherwise."""
+
+    def _resolve_parameters_(self: Any, param_resolver: ParamResolver):
+        """Resolve the parameters in the effect."""
+
+
+
+def is_parameterized(val: Any) -> bool:
+    """Returns whether a gate is parameterizable.
```
Comment (Strilanc) [inline cirq/protocols/resolve_parameters.py]: `parameterizable` is not the same as `parameterized`.

Also the object may not be a gate. Or could be an operation, or anything really.

File: cirq/protocols/resolve_parameters.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        return result
+    else:
+        return False
+
+
+def resolve_parameters(val: Any, param_resolver: ParamResolver) -> Any:
+    """Resolve the parameters in the effect.
+
+    Args:
+        val: The gate to resolve
```
Comment (Strilanc) [inline cirq/protocols/resolve_parameters.py]: May not be a gate.

File: cirq/protocols/resolve_parameters.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+    if result is not NotImplemented:
+        return result
+    else:
+        return False
+
+
+def resolve_parameters(val: Any, param_resolver: ParamResolver) -> Any:
+    """Resolve the parameters in the effect.
+
```
Comment (Strilanc) [inline cirq/protocols/resolve_parameters.py]: Specify what the magic method called by this method is.

File: cirq/protocols/resolve_parameters.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    Args:
+        val: The gate to resolve
+        param_resolver: the object to use for resolving all symbols
+
+    Returns:
+        a gate or operation of the same type, but with all Symbols
+        replaced with floats according to the given ParamResolver.
+        If no implementation exists, this will return the input
```
Comment (Strilanc) [inline cirq/protocols/resolve_parameters.py]: Specify what it means for no implementation to exist.

File: cirq/protocols/resolve_parameters.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        and False otherwise."""
+
+    def _resolve_parameters_(self: Any, param_resolver: ParamResolver):
+        """Resolve the parameters in the effect."""
+
+
+
+def is_parameterized(val: Any) -> bool:
+    """Returns whether a gate is parameterizable.
+
```
Comment (Strilanc) [inline cirq/protocols/resolve_parameters.py]: Specify the magic method being called by this method, and the default behavior when it doesn't exist or returns NotImplemented.

File: cirq/protocols/resolve_parameters.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    """An object that can be parameterized by Symbols and resolved
+    via a ParamResolver"""
+
+    def _is_parameterized_(self: Any) -> bool:
+        """Whether the gate is parameterized by any Symbols that require
+        resolution.  Returns True if the gate has any unresolved Symbols
+        and False otherwise."""
+
+    def _resolve_parameters_(self: Any, param_resolver: ParamResolver):
+        """Resolve the parameters in the effect."""
```
Comment (Strilanc) [inline cirq/protocols/resolve_parameters.py]: `Returns a copy of the given value, but with symbols replaced by concrete values from the given parameter resolver.`

File: cirq/protocols/resolve_parameters.py (hunk 5 above)
Comment (dstrain115) [reply cirq/protocols/resolve_parameters.py]: Ok, adjusted doc.  Let me know if you have a better phrasing.

File: cirq/protocols/resolve_parameters.py [hunk 11]
```diff
@@ -49,17 +52,20 @@ def is_parameterized(val: Any) -> bool:
 
 
 def resolve_parameters(val: Any, param_resolver: ParamResolver) -> Any:
-    """Resolve the parameters in the effect.
+    """Resolve the parameters in the effect by returning a copy of the
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply cirq/protocols/resolve_parameters.py]: Adjusted documentation.  Let me know if you would like alternate phrasing.

File: cirq/circuits/circuit.py [hunk 12]
```diff
@@ -27,6 +27,7 @@
 )
 
 import numpy as np
+import cirq
```
Comment (Strilanc) [inline cirq/circuits/circuit.py]: `from cirq import protocols`

File: cirq/circuits/circuit.py [hunk 13]
```diff
@@ -1210,14 +1206,7 @@ def _resolve_operations(
...
     for op in operations:
-        cast_op = ext.try_cast(ops.ParameterizableEffect, op)
-        if cast_op is None:
-            resolved_op = op
-        else:
-            resolved_op = cast(
-                    ops.Operation,
-                    cast_op.with_parameters_resolved_by(param_resolver))
-        resolved_operations.append(resolved_op)
+        resolved_operations.append(cirq.resolve_parameters(op, param_resolver))
```
Comment (Strilanc) [inline cirq/circuits/circuit.py]: Does this make `ext` a dead parameter? If so, delete it (and update calling methods to not pass it in, and delete ext from those methods if it becomes dead for them also)

File: cirq/protocols/resolve_parameters.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+    def _resolve_parameters_(self: Any, param_resolver: ParamResolver):
+        """Resolve the parameters in the effect."""
+
+
+
+def is_parameterized(val: Any) -> bool:
+    """Returns whether the object is parameterized with any Symbols.
+
+    A value is parameterized when it has an is_parameterized method and
```
Comment (Strilanc) [inline cirq/protocols/resolve_parameters.py]:
```
an `_is_parameterized_` method
```

The underscores are extremely important to include when referring to the method.

**Comment by dstrain115** [comment]
I think the Build failed due to some transitory error.  It doesn't look relevant to anything I did.  Is there a way to re-run it?

**Pull request merged**